### PR TITLE
[llvm-mc] Add --hex to disassemble hex bytes

### PR DESCRIPTION
**Pull request by MaskRay**
`--disassemble`/`--cdis` parses input bytes as decimal, 0bbin, 0ooct, or
0xhex. While the hexadecimal digit form is most commonly used, requiring
a 0x prefix for each byte (`0x48 0x29 0xc3`) is cumbersome.

Tools like xxd -p and rz-asm use a plain hex dump form without the 0x
prefix or space separator. This patch adds --hex to disassemble such hex
bytes with optional whitespace.

```
% rz-asm -a x86 -b 64 -d 4829c34829c4
sub rbx, rax
sub rsp, rax

% llvm-mc -triple=x86_64 --cdis --hex --output-asm-variant=1 <<< 4829c34829c4
        .text
        sub     rbx, rax
        sub     rsp, rax
```

### CODE REVIEW

**Comment by s-barannikov** [comment]
Maybe require hex pairs to be space separated? Otherwise a sequence of pairs looks like a single hex number and may cause confusion regarding endianness.

**Comment by MaskRay** [comment]
> Maybe require hex pairs to be space separated? Otherwise a sequence of pairs looks like a single hex number and may cause confusion regarding endianness.

Thanks for the quick response. Actually, it's intentional to make spaces optional. The hex pair form is used by tools like `xxd -p` (`xxd -p -r`) and rz-asm.

To interpret 0x12345678 as 0x78 0x56 0x34 0x12, we can keep using the existing functionality, or add another option if there is sufficient demand (I personally I don't think it is useful; easily misused as the length is not clear).

File: llvm/docs/CommandGuide/llvm-mc.rst [hunk 1]
```diff
@@ -92,6 +92,11 @@ End-user Options
 
  Generate DWARF debugging info for assembly source files.
 
+.. option:: --hex
+
+ Take hex pairs as input for the disassembler.
+ Whitespace is ignored.
+
```
Comment (s-barannikov) [inline llvm/docs/CommandGuide/llvm-mc.rst:99]: It looks like "hexpair" is a term invented by radare/rizin and is not used anywhere else.
"hex pair" (with a space) is even less popular, google shows less than ten pages.
How about just "hex bytes"?

(I'm not nitpicking, I actually had to look at the implementation to understand what it means.)

File: llvm/tools/llvm-mc/Disassembler.h [hunk 2]
```diff
@@ -32,7 +32,7 @@ class Disassembler {
   static int disassemble(const Target &T, const std::string &Triple,
                          MCSubtargetInfo &STI, MCStreamer &Streamer,
                          MemoryBuffer &Buffer, SourceMgr &SM, MCContext &Ctx,
-                         const MCTargetOptions &MCOptions);
+                         const MCTargetOptions &MCOptions, bool HexPairs);
```
Comment (s-barannikov) [inline llvm/tools/llvm-mc/Disassembler.h:35]:
```suggestion
                         const MCTargetOptions &MCOptions, bool HexBytes);
```

**Review by s-barannikov — APPROVED** [review]
LGTM

File: llvm/docs/CommandGuide/llvm-mc.rst (hunk 1 above)
Comment (s-barannikov) [reply llvm/docs/CommandGuide/llvm-mc.rst:99]: Thanks

**Comment by boomanaiden154** [comment]
Thanks very much for putting this together! I've been wanting a feature like this for quite a while for quick manual inspection of machine code serialized to hex.

**Pull request merged**